### PR TITLE
Open links when shift-click on link or url

### PR DESCRIPTION
**Pull request by ayaankazerouni**
> Before submitting your PR please make sure:
> * [x] You worked on the develop branch (or any other branch which was forked from develop),
> * [x] Your PR is not targeting the master branch,
> * [x] You tested your code and it works.

Related to: #198 

## Caveats:
The links only open when they're in the form `https://github.com`. Simple `github.com` or `www.github.com` won't do anything. I'm unsure if this is something electron does on purpose. @brrd Any ideas?

### CODE REVIEW

**Comment by brrd** [comment]
> Any ideas?

According to the [documentation about `shell.openExternal()`](https://electron.atom.io/docs/api/shell/#shellopenexternalurl-options-callback):

> Open the given external protocol URL in the desktop’s default manner. (For example, mailto: URLs in the user’s default mail agent).

So I think if the URL doesn't start with the protocol then the shell module can not guess how to open it.

Why not adding it when it misses?

File: app/renderer/abr-document.js [hunk 1]
```diff
@@ -207,6 +216,35 @@ function AbrDocument () {
...
+          var range = that.cm.getRange(word.anchor, word.head);
+          var open = null;
+          if (e.target.classList.contains("cm-url")) {
+              open = $(e.target).prevAll(".cm-formatting-link-string").first();
+          } else if (e.target.classList.contains("cm-link")) {
+              open = $(e.target).nextAll(".cm-formatting-link-string").first();
+          }
+
+          var url = open.nextUntil(".cm-formatting-link-string").text();
+          if (url !== "") {
```
Comment (brrd) [inline app/renderer/abr-document.js]:
```javascript
if (url === "") return;
var hasProtocol = /^[a-z]+:\/\//.test(url);
if (!hasProtocol) {
    url = "http://" + url;
}
shell.openExternal(url);
```

**Comment by brrd** [comment]
Few comments about this PR:

**Raw URL**

It obviously doesn't work with raw URL:

```markdown
This is working: [Test](http://google.com)

But clicking this does nothing: http://google.com
```

**Pointer**

I don't see the pointer when Alt key is pressed on a link, but a crosshair (I tried it on Windows 10). Probably related to https://github.com/codemirror/CodeMirror/blob/dda5d7220f0e3cbf8c203223ee4e7dc2d5c04bec/src/edit/key_events.js#L119

**Alt key**

I would not use Alt key because 1) on Windows and some Linux DE the Alt key also triggers the menu if it's hidden, making the editor contents move with possible confusion about what is clicked, 2) Alt is already used by CodeMirror for column selection. 

Control/command is already used for multiple cursors and Shift is used for extending the current selection. My suggestion would be to:

1. use the Shift key 
2. trigger the function on "mouseup" (instead of "mousedown")
3. stop execution if something is selected in the editor (`cm.doc.somethingSelected()`)

This would also solve the pointer issue.

What do you think?

**Comment by ayaankazerouni** [comment]
> Raw URL

Ok, this is dealt with. The following cases work:
```markdown
http://google.com (raw url)
http://github.com ("github" is seen as a spelling error, so it breaks the link into `spans`)
[Facebook](facebook.com) (standard MD link without protocol)
[Facebook](https://facebook.com) (standard MD link)
```

> Pointer, Alt Key

Ok, using Shift key now, when nothing is selected. CodeMirror doesn't seem to emit a `mouseup` event. What are the downsides of using `mousedown`?

**Comment by brrd** [comment]
> What are the downsides of using mousedown?

Selection is not updated yet when `mousedown` event is fired, so `cm.somethingSelected()` will refer to the previous selection.

> CodeMirror doesn't seem to emit a mouseup event.

You don't need to use CodeMirror events, you can simply do:

```javascript
document.addEventListener("mouseup", function () {
// Do stuff
});
```

File: app/renderer/abr-document.js [hunk 2]
```diff
@@ -207,6 +218,48 @@ function AbrDocument () {
             }
         };
 
+        // Handle ALT modifier key
+        var ShiftKeyHandler = function(e) {
+            var linkIsClickable = !that.cm.somethingSelected() &&
+                e.type === "keydown" &&
+                e.shiftKey;
+            document.body.classList.toggle("shift-pressed", linkIsClickable);
```
Comment (brrd) [inline app/renderer/abr-document.js]: This part is smart, but I had to read it twice to understand why you did it this way. Could you rename the class name in something less confusing (maybe "link-clickable")?

**Comment by ayaankazerouni** [comment]
I now open the link on `mouseup` as requested. But I have to listen to `mousedown` event to `preventDefault` the standard shift-click behaviour (it selects text between the original cursor position and the clicked position). I check for the same set of conditions as in `mouseup` to avoid to preventing default behaviour any more than necessary.

**Comment by brrd** [comment]
Thanks!

**Comment by ayaankazerouni** [comment]
No problem! This is a great app! Do you have an ETA on the next release?

**Comment by brrd** [comment]
As soon as possible! (I'm too busy right now 😕)

**Pull request merged**